File: LeetCode Practice Problems/39. Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> ans;
    void find(vector<int> a,vector<int> path,int index,int target)
    {
        if(target < 0) return;
        if(target == 0)
        {
            ans.push_back(path);
            return;
        }
        
        for(int i=index;i<a.size();i++)
        {
            int v = a[i];
            path.push_back(v);
            find(a,path,i,target - v);
            path.pop_back();
        }
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> p;
        find(candidates , p,0,target);
        return ans;
    }
};

File: LeetCode/LeetCode Practice Problems/39-Combination Sum.java
class Solution {
    
    private void backtrack(List<List<Integer>> ans,List<Integer> res,int[] nums,int remain,int start) {
        if(remain < 0) return;
        if(remain == 0) {
            ans.add(new ArrayList<>(res));
            return;
        }
        for(int i=start;i<nums.length;i++) {
            res.add(nums[i]);
            backtrack(ans,res,nums,remain - nums[i],i);
            res.remove(res.size() - 1);
        }
    }
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList();
        Arrays.sort(candidates);
        backtrack(ans,new ArrayList(),candidates,target,0);
        return ans;
    }
}
